redis/redis.go

```diff
@@ -86,7 +112,8 @@ func (b *bucket) Add(amount uint) (leakybucket.BucketState, error) {
 	b.updateOldReset()
 
 	// Ensure we can't overflow
-	b.remaining = b.capacity - min(uint(count), b.capacity)
+	x := b.Capacity()
+	b.setRemaining(x - min(uint(count), x))
 	return b.State(), nil
```

Comment:
Did you consider locking at the start of Add()? As it reads now, there is a lot of locking/unlocking happening within the method via the `set*` methods, and it's hard to reason about the possible access patterns.